File: strings/semordnilap.js
// Write a function semordnilap() that takes a list of unique strings and returns a list of semordnilap pairs.
// A semordnilap is a word or a phrase that spells a different word when backwards ("semordnilap" is a semordnilap of "palindromes").

// My solution 9/25/23
function semordnilap(words) {
  let copyWords = JSON.parse(JSON.stringify(words));

  const resultArr = [];

  while (copyWords.length) {
    let currentWord = copyWords.shift();
    let reverseWord = currentWord.split("").reverse().join("");
    for (let i = 0; i < copyWords.length; i++) {
      if (copyWords[i] === reverseWord) {
        let duple = [];
        duple.push(currentWord, reverseWord);
        resultArr.push(duple);
        copyWords.splice(i, 1);
      }
    }
  }
  return resultArr;
}

File: strings/semordnilap.py
# Write a function semordnilap() that takes a list of unique strings and returns a list of semordnilap pairs.
# AlgoExpert version:

def semordnilap(words):
    # This line creates a set from the list of words. A set is an unordered collection of unique elements. By converting the list to a set, it removes any duplicate words in the input list, which simplifies the process of finding semordnilaps.
    # The use of a set makes the code more efficient when checking for semordnilaps since set membership checks are typically faster than list membership checks, especially for larger datasets. Additionally, it ensures that each word is considered only once, even if it appears multiple times in the input list.
    wordsSet = set(words)
    semordnilapPairs = []

    for word in words:
        reversedWord = word[::-1]
        if reversedWord in wordsSet and reversedWord != word:
            semordnilapPairs.append((word, reversedWord))
            wordsSet.remove(reversedWord)
            wordsSet.remove(word)

    return semordnilapPairs

print(semordnilap(['live', 'evil', 'live', 'evil', 'dog', 'god', 'cat', 'tac']))
# [('evil', 'live'), ('god', 'dog'), ('tac', 'cat')]
